StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/PaymentSheetDeferredValidator.swift

```diff
@@ -0,0 +1,67 @@
+//
+//  PaymentSheetDeferredValidator.swift
+//  StripePaymentSheet
+//
+//  Created by Yuki Tokuhiro on 5/16/23.
+//
+
+import Foundation
+import StripePayments
+
+struct PaymentSheetDeferredValidator {
+    static func validate(paymentIntent: STPPaymentIntent, intentConfiguration: PaymentSheet.IntentConfiguration) throws {
+        guard case let .payment(amount, currency, setupFutureUsage, captureMethod) = intentConfiguration.mode else {
+            throw PaymentSheetError.unknown(debugDescription: "You returned a PaymentIntent client secret but used a PaymentSheet.IntentConfiguration in setup mode.")
+        }
+        guard paymentIntent.currency.uppercased() == currency.uppercased() else {
+            throw PaymentSheetError.unknown(debugDescription: "Your PaymentIntent currency (\(paymentIntent.currency)) does not match the PaymentSheet.IntentConfiguration currency (\(currency)).")
+        }
+        guard paymentIntent.setupFutureUsage == setupFutureUsage else {
+            throw PaymentSheetError.unknown(debugDescription: "Your PaymentIntent setupFutureUsage (\(paymentIntent.setupFutureUsage)) does not match the PaymentSheet.IntentConfiguration setupFutureUsage (\(String(describing: setupFutureUsage))).")
+        }
+        guard paymentIntent.amount == amount else {
+            throw PaymentSheetError.unknown(debugDescription: "Your PaymentIntent amount (\(paymentIntent.amount)) does not match the PaymentSheet.IntentConfiguration amount (\(amount)).")
+        }
+        guard paymentIntent.captureMethod == captureMethod else {
+            throw PaymentSheetError.unknown(debugDescription: "Your PaymentIntent captureMethod (\(paymentIntent.captureMethod)) does not match the PaymentSheet.IntentConfiguration amount (\(captureMethod)).")
+        }
+    }
+
+    static func validate(setupIntent: STPSetupIntent, intentConfiguration: PaymentSheet.IntentConfiguration) throws {
+        guard case let .setup(_, setupFutureUsage) = intentConfiguration.mode else {
```

Comment:
Why not validating currency for setup but we do for payment?

Reply:
SetupIntent doesn't have a currency, so we can't validate.

Reply:
Oh true.. I saw that our intent config mode Setup has a currency but forgot we only use that for UI formatting so got mixed up.